tests/foreman/cli/test_leapp_client.py

```diff
@@ -0,0 +1,220 @@
+    # rhel_contenthost = custom_host
+    organization = setup_env['organization']
+    target_sat = setup_env['satellite']
+    lc_env = setup_env['environment']
+    c_view = setup_env['contentview']
+
+    # Enable rhel8/rhel9 bos, aps repository and add in content view
+    all_repos = []
+    for rh_repo_key in CUSTOM_REPOS.keys():
+        prod = rh_repo_key.split('_')[0]
+        repo_id = target_sat.api_factory.enable_rhrepo_and_fetchid(
+            basearch=custom_host.arch,
+            org_id=organization.id,
+            product=PRDS[prod],
+            repo=CUSTOM_REPOS[rh_repo_key]['name'],
+            reposet=CUSTOM_REPOSET[rh_repo_key],
+            releasever=CUSTOM_REPOS[rh_repo_key]['releasever'],
+        )
+        rh_repo = target_sat.api.Repository(id=repo_id).read()
+        all_repos.append(rh_repo)
+        # sync repo
+        rh_repo.sync(timeout=1800)
+    c_view.repository = all_repos
+    c_view = c_view.update(['repository'])
+    # Publish, promote content view to lce
+    c_view.publish()
+    cv_version = c_view.read().version[0]
+    cv_version.promote(data={'environment_ids': lc_env.id, 'force': True})
+    c_view = c_view.read()
+    # Create activation key
+    ak = target_sat.api.ActivationKey(
+        content_view=c_view,
+        environment=lc_env,
+        organization=organization,
+    ).create()
+    # 3. Register Host
+    custom_host.install_katello_ca(target_sat)
+    custom_host.register_contenthost(org=organization.label, activation_key=ak.name)
+    custom_host.add_rex_key(satellite=target_sat)
+    assert custom_host.subscribed
+    # 4. Verify target rhel version repositories has enabled on Satellite Server
+    cmd_out = target_sat.execute(
+        f"hammer repository list --search 'content_label ~ rhel-9' --content-view {c_view.name} "
+        f"--organization '{organization.name}' --lifecycle-environment '{lc_env.name}'"
+    )
+    assert cmd_out.status == 0
+    assert ("AppStream RPMs 9" in cmd_out.stdout) and ("BaseOS RPMs 9" in cmd_out.stdout)
+
+    # Preupgrade conditions and check
+    custom_host.run("rm -rf /root/tmp_leapp_py3")
+    # Remove directory if in-place upgrade already performed from RHEL7 to RHEL8
+    rhel_old_ver = custom_host.run('cat /etc/redhat-release')
+    # 5. Update all packages and install Leapp utility
+    custom_host.run("dnf clean all")
+    custom_host.run("dnf repolist")
+
+    custom_host.run("subscription-manager release --set 8.8")
+    result = custom_host.run("dnf update -y")
+    assert result.status == 0
+    custom_host.run("dnf install leapp-upgrade -y")
+
+    # Fixing inhibitors - download data files to avoid inhibitors
+    custom_host.run(
+        'sed -i "s/^AllowZoneDrifting=.*/AllowZoneDrifting=no/" /etc/firewalld/firewalld.conf'
+    )
+    if custom_host.run('needs-restarting -r').status == 1:
+        custom_host.power_control(state='reboot', ensure=True)
+        custom_host.power_control(state=VmState.RUNNING, ensure=True)
```

Comment:
Can you explain why you need to run power_control twice?

Generally, if you reboot the system, it should boot up again, therefore you should not need to ensure that it is really running.

Reply:
Okay, I think, right after reboot finish with `power_control` at #167, the script gets back ssh connection and no need to use `power_control` again because `wait_for` already inplace inside.
@ogajduse Am I assuming correctly?

Reply:
Cleared.